### ISSUE TITLE
Debugger calls getters to update variable inspector

### ISSUE THREAD

**Issue by IndustrialDonut** [godotengine/godot#50967]
### Godot version

3.3.2

### System information

Windows 10

### Issue description

The actual debugger itself calls getters when they are defined for a variable, so as to update the debugger variable inspector. It should not do this, for this clearly can alter the state of the program from what it would be without the debugger, and introduce new issues solely from the debugger that are not the real issues in the first place.

### Steps to reproduce

Open the project and the only Node2D scene in there, and run the scene. Note that 'times_called' starts at 0, and never anywhere at all is 'variable' referenced, so the getter should never be called. Put a breakpoint on line 14 and step down one or two lines, and look at times_called in the inspector. Then, remove the breakpoint and run the scene. I get something like 48 printing for 'times_called' after using debugger to step a couple of lines, vs 0 when I run without the breakpoint.

### Minimal reproduction project

[debugger bug.zip](https://github.com/godotengine/godot/files/6891190/debugger.bug.zip)

**Comment by hilfazer** [comment]
Whether it's a bug or not i have no idea but this is useful for specifying the way our custom types get displayed in the Inspector during debugging.

I had classes that implement the behaviour of set and map data structures and used custom getter to display them in the inspector in a more approachable way - instead of displaying an Object i was displaying an underlying container (Array or Dictionary).

**Comment by KoBeWi** [comment]
This is most likely intended, the getter might return a completely different value than the actual variable.

Getters should be const. If you change the state of the program inside getter it's a logic flaw.

**Comment by IndustrialDonut** [comment]
> This is most likely intended, the getter might return a completely different value than the actual variable.
> 
> Getters should be const. If you change the state of the program inside getter it's a logic flaw.

This is another reason why it doesn't make sense for it to be intended though, the variable inspector should be inspecting exactly that, the value of the variable. It should not be inspecting the getter, which could in theory be totally different.

**Comment by SpaceAceMonkey** [comment]
I was coming here to report this same issue. I don't see how this is desirable behavior. There are plenty of reasons you might want to do things inside your getter that change state. I frequently use getters to load a resource if it hasn't yet been loaded so I don't have to litter my code with checks to determine whether or not the resource is null, is the right type, and so on.

If we really wanted to live in a world with no side-effects, we'd be writing all our code in purely functional languages. I don't think "you shouldn't do that" is a good argument when "that" is something extremely common that nearly all languages allow and every programmer I've ever known has made use of.

**Comment by Ev1lbl0w** [comment]
After some quick testing I found issue is not caused by the debugger, but by the inspector instead. If you change the remote inspector interval to something big like every 100 seconds, you'll see the `times_called` on the debugger stacktrace is no longer incrementing like crazy:
![Peek 2021-09-16 21-36](https://user-images.githubusercontent.com/6501975/133682351-051b9a43-2247-4ec5-9680-ad59dbdd4163.gif)

Maybe this behavior should be changed when using the remote inspector to retrieve the value of members directly like the debugger does, instead of going through getters *(and keeping current behavior for local inspector to not break `tool` scripts relying on `setget`)*

**Comment by tx350z** [comment]
There are numerous questions and issues concerning property access modifiers (private and protected). In most cases the advice is to use setget methods to control access to these properties. However, setget methods create problems when debugging. Consider the following example where __crypto_key needs to be accessible only from within the defining class and descendant classes and the debugger of course.

```
var __crypto_key:CryptoKey setget _set_crypto_key, _get_crypto_key;
func _set_crypto_key(_val:CryptoKey):
	push_error("Attempt to set protected property '__crypto_key' prohibited.")
func _get_crypto_key():
	push_error("Attempt to get protected property '__crypto_key' prohibited. Call get_crypto_key() instead.")
func get_crypto_key()->String:
	return __crypto_key.save_to_string(true);

```
When run in the debugger the errors are pushed repeatedly. So either we need a way for the debugger to access properties without calling the getter method (e.g. no erroneous errors) OR we need access modifiers that the debugger can ignore.

**Comment by FeralBytes** [comment]
Further I have to ask how would one suggest that we work around this issue?

**Comment by rsubtil** [comment]
> Yes semantics it is the inspector causing this and not the debugger, but it is specifically remotely inspecting for the purpose of debugging. Unfortunately it is causing bugs.

Well, the issue with this being in the Inspector is that there can be legitimate use cases for it going through a getter instead of accessing the object itself, like @hilfazer pointed out. But likewise, there are use cases where the current behavior is problematic, like @tx350z's example. 

Can you describe a bit your use-case for a mutating getter? I'm having a hard time finding good examples of such cases, and it feels like an extremely problematic decision, especially for debugging future issues in your project.

If you're doing something similar to OP's code and counting how many times a getter is called, you should use [the profiler](https://docs.godotengine.org/en/stable/tutorials/scripting/debug/the_profiler.html) for that.

**Comment by FeralBytes** [comment]
@rsubtil Yes. My use case was for providing each client with a persistent identification number that is handed out by the server the starting id that is given to the server is not known in advanced as a it is a random range. Clients can then disconnect and re-connect and the server still knows them from the previous connection because they present this persistent id as part of being validated. 
I do see, now, that there are also legitimate uses for wanting the inspector to use the getter. I suppose since both use cases are valid it would be nice in the future to have a want to annotate that a variable should not be accessed via the getter. I doubt that any such flag will not be implemented for a long while yet.
I will have to try using the profiler as you describe. Thank you for pointing that out. I also have changed my code to a safer technique by using a custom function to get the next persistent id in the servers code.
